Fix #9733.  Add valgrind documentation and suppressions

garrison opened the pull request:
Here is some documentation for using Valgrind with Julia.  With regards to the "target architecture mismatch" error, I decided to describe the status quo rather than attempt to disable the check when `RUNNING_ON_VALGRIND` is true (as suggested by @Keno).  This can always be changed once the current pull request is merged.  (I figured it may be helpful to describe the status quo in case this documentation is e.g. backported to the `release-0.3` branch.)

garrison committed:
Fix #9733.  Add valgrind documentation and suppressions

timholy commented:
I am not remotely a valgrind guru, but I wonder particularly about the `fun:rec_backtrace_ctx` in your suppression file. At one point I noticed a genuine memory leak in our backtraces, but I haven't had time to fix it yet. Not sure if someone else has already done so. Anyway, just wanting to raise the issue in case there's any risk you're suppressing a genuine problem that we should just deal with.

garrison commented:
Not sure if you saw the coment, but the goal there is to suppress the issue reported in #4533.

timholy commented:
Ah, I understand better now. Thanks.

garrison committed:
Remove valgrind doc section on arch mismatch

Now that #9985 is merged, this section is irrelevant for julia master (0.4)

[ci skip]

garrison commented:
I think this is ready to go.  @Keno, do you have any further thoughts?

Keno commented:
LGTM
